scrapy/exporters.py

```diff
@@ -252,13 +258,21 @@ def _build_row(self, values):
 
     def _write_headers_and_set_fields_to_export(self, item):
         if self.include_headers_line:
+            if isinstance(item, dict):
+                # for dicts try using fields of the first item
+                fields = list(item.keys())
+            else:
+                # use fields declared in Item
+                fields = list(item.fields.keys())
             if not self.fields_to_export:
-                if isinstance(item, dict):
-                    # for dicts try using fields of the first item
-                    self.fields_to_export = list(item.keys())
-                else:
-                    # use fields declared in Item
-                    self.fields_to_export = list(item.fields.keys())
+                self._is_data_loss_warning_displayed = True
+                self.fields_to_export = fields
+            elif not self._fail_on_dataloss_warned:
+                if fields != self.fields_to_export:
+                    self._is_data_loss_warning_displayed = True
+            if self._is_data_loss_warning_displayed and not self._fail_on_dataloss_warned:
+                logger.warning("Possible chance of data loss detected -- This message won't be shown in further requests")
```

Comment:
I think this warning should:
* Appear only when there is actual data loss, i.e. when the current fields are definitely different from the stored in the exporter instance. Hence "Possible chance of" should go away IMHO (https://github.com/scrapy/scrapy/issues/4002#issuecomment-536662554)
* Give a little bit more information about the data loss, maybe "Data loss detected **when exporting items**" (although this might be covered by the logger name). Perhaps mentioning that some fields are not being included in the output.
* Talk about items instead of requests ("This message won't be shown **for further items**")

Reply:
Cool I will do those fixes, but what if `fields_to_export` is not set but there is no actual data loss.In other words the row size and fields remain the same for all fields but the `fields_to_export` is not set. Do I need to show the same message?

Reply:
I think there should be no warning in that case (https://github.com/scrapy/scrapy/pull/4053#discussion_r331736731)